Automatic scheduling directives

tdenniston opened the pull request:
Hello,

This PR adds the notion of "optimization levels" for some simple automatic scheduling. The idea here is to offer better default schedules to the programmer. To use these default schedules, set the new `HL_SCHED_OPT` environment variable to 0, 1, or 2:
- `HL_SCHED_OPT=0` is the default, which makes no changes to the schedule. This matches the current behavior.
- `HL_SCHED_OPT=1` recursively schedules functions from the pipeline output according to the following heuristic: functions with a single callsite are inlined; functions called as a stencil are `compute_root`.
- `HL_SCHED_OPT=2` first applies level 1, and then for each `compute_root` function, the outer dimension is parallelized and the inner dimension is vectorized (with a function type dependent vectorization factor).

Higher levels can be implemented later (I have one in the pipeline).

Currently, optimizations level 1 and 2 reset all function schedules before running the heuristics. That means if the user manually schedules some portion of the pipeline, but requests one of these automatic schedules, the user-specified schedule changes are discarded.

This is my first foray into modifying Halide, so I'm sure there are changes that must be made before this can be entertained for merging.

zvookin commented:
Thank you for the contribution. I'd like this a lot more it were driven by a scheduling directive, not an environment variable. Vectorization imposes a constraint that the extent of the dimension vectorized be at least as large as the vectorization factor, which can result in a change of correctness for code when it affects inputs or outputs. Driving such a change off an environment variable is really bad. In general, we should kill off environment variable usage, but it is very convenient when doing small hacks and tests. But I'd like to avoid adding more and adding a feature that is only available via environment variables is a very bad idea.

If you have any benchmark data on how well this scheduling works in practice I'd be interested. Also, did you consider any other ideas for easier, but still somewhat explicit, scheduling?

abadams commented:
To expand on making it a scheduling directive: something like
func.auto_schedule(some descriptive enum) could be good. The enum could be
something like ComputeRootAllStencils, or it could be a more descriptive
set of parameters. We needn't slavishly stick to a numbering of the
different strategies.

On Tue, Mar 24, 2015 at 12:26 PM, Zalman Stern <notifications@github.com>
wrote:

> Thank you for the contribution. I'd like this a lot more it were driven by
> a scheduling directive, not an environment variable. Vectorization imposes
> a constraint that the extent of the dimension vectorized be at least as
> large as the vectorization factor, which can result in a change of
> correctness for code when it affects inputs or outputs. Driving such a
> change off an environment variable is really bad. In general, we should
> kill off environment variable usage, but it is very convenient when doing
> small hacks and tests. But I'd like to avoid adding more and adding a
> feature that is only available via environment variables is a very bad idea.
>
> If you have any benchmark data on how well this scheduling works in
> practice I'd be interested. Also, did you consider any other ideas for
> easier, but still somewhat explicit, scheduling?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/712#issuecomment-85658267>.
>

zvookin commented:
The argument to the scheduling directive can be provided by a GeneratorParam, which allows for enums.

tdenniston commented:
Thanks for the feedback. I'll get to work on implementing this as a scheduling directive instead. In the meantime, here are some performance numbers. I ran a few of the programs in the 'apps' directory on a machine with (12 core) Intel Xeon E5-2695 v2 @ 2.40GHz. The "manual" schedule is what is in the repository. I ran each benchmark 5 times in a row and took an average.

* blur (manual)
avg times: 5.1532334	0.1278588	0.0510634

* blur (opt 1)
avg times: 5.10596	0.112125	2.0005608

* blur (opt 2)
avg times: 5.0777346	0.1184314	0.4440868

* bilateral_grid (manual)
avg: 10.07094ms

* bilateral_grid (opt 1)
avg: 126.6087ms

* bilateral_grid (opt 2)
avg: 77.10564ms

* camera_pipe (manual)
Halide avg: 6443.8
C++: avg: 112565.8

* camera_pipe (opt 1)
avg: 130957

* camera_pipe (opt 2)
avg: 34436

* local_laplacian (manual)
avg: 153067

* local_laplacian (opt 1)
avg: 896059

* local_laplacian (opt 2)
avg: 513951

tdenniston commented:
Also, I'm not sure what to do as far as naming these scheduling strategies goes. `ComputeRootAllStencils` is clear, but what should the "level 2" be named? `ComputeRootAllStencilsParallelizeVectorize` seems like a bit of a mouthful. And my eventual level 3 would be something even longer.

One alternative is to construct the different strategies such that they compose. Level 2 would then be the composition of `ComputeRootAllStencils` and `ParallelizeOuterVectorizeInner`. The programmer would have to specify both strategies explicitly.

abestephensg commented:
"Auto-schedules" or "scheduling patterns" make sense to me if it is clear that they are a starting place for a developer to tune the schedule.  The example above shows that it is not a short cut to avoid tuning:  "opt 1" for bilateral grid is 12x slower. The mechanism should make it clear that this is a schedule variant and the developer must still test different variants.

The alternative of constructing strategies or patterns from several directives helps do this. In some sense it becomes a form of syntactic sugar that allows the developer to apply a scheduling policy broadly across a top level Func.  Instead of keeping track of individual lower level Funcs and calling `compute_root` on them, you can call `.auto_schedule(ComputeRootAllStencils)` or similar.

tdenniston commented:
Structuring it this way might lead to some complications though. For example, suppose we created two new strategies that did the following: 1) tile functions under some circumstances; 2) for a function `f`, if its caller is tiled, compute and store `f` at tile granularity. 

Under this design, how would the second strategy examine a function to see if its caller was tiled? If these strategies were fused into one, it could simply record which functions it tiled and use that in the later check for function `f`. But with the design we are discussing, the two strategies would be implemented independently, and I don't know how they could communicate in this way.

abadams commented:
We could certainly add ways to introspect schedules. However, I think Abe
was just thinking of cases like:

output.auto_schedule(ComputeRootAllStencils);
output.bound(c, 0, 3).unroll(c);

where the information you're delivering is largely independent of the
decisions made by the auto scheduler.



On Wed, Mar 25, 2015 at 11:46 AM, Tyler Denniston <notifications@github.com>
wrote:

> Structuring it this way might lead to some complications though. For
> example, suppose we created two new strategies that did the following: 1)
> tile functions under some circumstances; 2) for a function f, if its
> caller is tiled, compute and store f at tile granularity.
>
> Under this design, how would the second strategy examine a function to see
> if its caller was tiled? If these strategies were fused into one, it could
> simply record which functions it tiled and use that in the later check for
> function f. But with the design we are discussing, the two strategies
> would be implemented independently, and I don't know how they could
> communicate in this way.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/712#issuecomment-86168381>.
>

tdenniston commented:
Ok, I've implemented this as a scheduling directive. Now you can do, e.g.

    bilateral_grid.auto_schedule(ComputeRootAllStencils)
      .auto_schedule(ParallelizeOuter)
      .auto_schedule(VectorizeInner);

The user-specified schedules are still reset with this implementation. I implemented it using a regular `enum` because I didn't find any examples of how to use the `GeneratorParam` as an enum.

Is this something that people will find useful?

steven-johnson commented:
test/generator/example_generator.cpp has an example of GeneratorParam as enum; see the SomeEnum type and the "enummy" member variable.

tdenniston commented:
@steven-johnson I did see that, but I was unable to find an example of how  `enummy` is actually used. It's only defined in that file, as far as I can see.

steven-johnson commented:
Oops. Bad example then. I'd better upgrade it. In the meantime: within the build() method, just reference it like any enum:

    if (enummy == Foo) { ... }

to set it, follow the same approach you'd use for any other generator argument, but use the string in the map, e.g.

    run_my_generator enummy=foo

dsharletg commented on Makefile:
```diff
@@ -214,6 +214,7 @@ FILTERS_DIR = $(BUILD_DIR)/filters
 
 SOURCE_FILES = \
   AllocationBoundsInference.cpp \
+  AutomaticScheduling.cpp \
```
Please also add the new files to the src/CMakeLists.txt build.

dsharletg commented on src/AutomaticScheduling.cpp:
```diff
@@ -0,0 +1,268 @@
+    internal_assert(impl != NULL);
+    // Reset all user-specified schedules.
+    if (reset_schedules) ResetSchedules reset(root.function());
+    impl->apply(root);
```
It looks like there's a leak of 'impl' here.

steven-johnson replied:
Please let's officially flip that bit; unique_ptr alone makes it all worthwhile :-)

abadams replied:
Consider it flipped. Probably also want to use range-based for loops for the iterators over vectors of Functions above.

dsharletg commented on src/AutomaticScheduling.h:
```diff
@@ -0,0 +1,64 @@
+/** Base class for all automatic scheduling strategy implementations. */  
+class AutoScheduleStrategyImpl {
+public:
+    /** Apply the schedule strategy to the pipeline. 'func' should
```
I know the inherited classes don't have any member variables yet, and the design is such that this is unlikely to be the case in the future. But it would probably still be a safe choice to add a virtual destructor, since we free these via the base class. It's really easy to forget adding this later.

dsharletg commented:
This is a cool change! I will use this often when I'm beginning work on a new algorithm and just need any schedule other than an inline explosion. I added a few comments to the code. 

Also, this PR needs to be updated before we can merge it with master.

dsharletg commented on src/AutomaticScheduling.cpp:
```diff
@@ -0,0 +1,268 @@
+    /** Return list of functions directly calling function f. */
+    const vector<Function> &callers(Function f) {
+        map<string, vector<Function> >::const_iterator I = call_to_caller.find(f.name());
+        if (I == call_to_caller.end()) {
```
I think this logic would be equivalent to simply:

    return call_to_caller[f.name()];

std::map's operator[] would simply return a default constructed vector<Function> if the key doesn't exist.

dsharletg commented on src/AutomaticScheduling.cpp:
```diff
@@ -0,0 +1,268 @@
+    /** Return list of functions directly called by f. */
+    const vector<Function> &calls(Function f) {
+        map<string, vector<Function> >::const_iterator I = call_to_callee.find(f.name());
+        if (I == call_to_callee.end()) {
```
Ditto the comment above regarding std::map's operator []

dsharletg commented on src/AutomaticScheduling.cpp:
```diff
@@ -0,0 +1,268 @@
+        Schedule &olds = f.schedule();
+        std::vector<Bound> oldbounds(olds.bounds().begin(), olds.bounds().end());
+        f.schedule() = Schedule();
+        Schedule &s = f.schedule();
```
I believe these two lines could be slightly more easily written:

    Schedule &s = f.schedule();
    s = Schedule();

dsharletg commented on src/AutomaticScheduling.cpp:
```diff
@@ -0,0 +1,268 @@
+    result.resize(f.dimensions());
+    vector<bool> initialized;
+    initialized.resize(f.dimensions(), false);
+    vector<Function> callers = cg.callers(f);
```
Could this be a reference/const reference to the vector in cg? I doubt its an issue now, but we've had problems in the past where passing things by value eventually added up to significant performance issues. Ditto this in a few more places below.

abadams commented on src/AutomaticScheduling.cpp:
```diff
@@ -0,0 +1,268 @@
+    CallGraph cg(root.function());
+    vector<Function> all_functions = cg.transitive_calls(root.function());
+    all_functions.push_back(root.function());
+    for (vector<Function>::iterator I = all_functions.begin(), E = all_functions.end(); I != E; ++I) {
```
Either use a range-based for loop for these loops over vectors, or just use:

for (size_t i = 0; i < all_functions.size(); i++) {
  Function f = all_functions[i];
  ...
}

which is the idiom used elsewhere in the Halide source.

abadams commented on src/Func.h:
```diff
@@ -414,6 +415,11 @@ class Func {
 
+    /** Whether or not function schedules should be reset before
+     * performing any automatic scheduling. */
+    bool auto_schedule_reset;
```
What's the motivation for ever resetting the user schedule? I could see some things getting accidentally silently dropped (e.g. .bound() directives)

tdenniston replied:
The original motivation for that was two things:
- I didn't know of a way to test if a function's schedule was modified or not. E.g. we could have the auto-scheduler only attempt to auto-schedule functions whose schedules have not been modified by the user.
- For the more powerful heuristic scheduler in the works, I wanted the scheduler to have full control, and I wanted to be sure that the performance numbers I saw were from the heuristic scheduler, not the manual schedule.

But neither of those seems to be a very good reason for it to be in this particular PR, so I could remove it if you think it's best.

tdenniston commented:
Ok, I addressed the comments here (except for the unresolved question of whether the ResetSchedules process should remain).

abadams commented:
After thinking about it, I believe schedules should not be reset. It'll
just trigger an error if some scheduling done before the auto schedule
conflicts with the decisions made by the auto-scheduler. This lets you do
things like change the notion of inner and outer for an individual Func if
you want a slight modification on the auto_schedule for that Func only:

Func f;
f(x, y, c) = foo;
f.reorder(x, c, y);

output_func.auto_schedule(VectorizeInner).auto_schedule(Parallelize_Outer);

With this in mind, vectorize inner should probably only vectorize the
innermost dimension that is not already unrolled, and should be a no-op if
it encounters a dimension already marked as vectorized.

The other major thing missing from this PR is something in test/correctness
that takes this stuff for a spin.


On Tue, Mar 31, 2015 at 9:47 AM, Tyler Denniston <notifications@github.com>
wrote:

> Ok, I addressed the comments here (except for the unresolved question of
> whether the ResetSchedules process should remain).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/712#issuecomment-88167917>.
>

abadams commented on src/AutomaticScheduling.cpp:
```diff
@@ -0,0 +1,255 @@
+            Func wrapper(f);
+            Dim inner = f.schedule().dims()[0];
+            Var v(inner.var);
+            unsigned factor = 128 / f.output_types()[0].bits;
```
This number should probably be 256 so that it hits avx. On targets with 128-bit vectors the backend should slice things up cleanly.

steven-johnson replied:
Do we have access to a Target at this point? It would be nice to use Target::natural_vector_size() instead.

tdenniston commented:
Ok, that seems like a reasonable plan to me. One thing I don't know how to implement is testing for conflicting schedules in general. The only approach that comes to mind is calling validate_schedule (or whatever it's called) after each individual decision is made. However, that seems like a slow approach.

I'll also get started on some correctness tests.

tdenniston commented:
To clarify: should the tests in tests/correctness only validate the output of a pipeline which has been auto-scheduled? Or should they validate that the resulting schedule after auto-scheduling is correct?

zvookin commented:
+1 to Steven's comment. 256 bit on AVX prior to AVX2 is a lose for integer
ops.

-Z-


On Tue, Mar 31, 2015 at 10:19 AM, Tyler Denniston <notifications@github.com>
wrote:

> To clarify: should the tests in tests/correctness only validate the output
> of a pipeline which has been auto-scheduled? Or should they validate that
> the resulting schedule after auto-scheduling is correct?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/712#issuecomment-88176981>.
>

abadams commented:
I guess auto_schedule could take a target - after all, the "correct"
schedule is probably target-specific.



On Tue, Mar 31, 2015 at 10:44 AM, Zalman Stern <notifications@github.com>
wrote:

> +1 to Steven's comment. 256 bit on AVX prior to AVX2 is a lose for integer
> ops.
>
> -Z-
>
>
> On Tue, Mar 31, 2015 at 10:19 AM, Tyler Denniston <
> notifications@github.com>
> wrote:
>
> > To clarify: should the tests in tests/correctness only validate the
> output
> > of a pipeline which has been auto-scheduled? Or should they validate that
> > the resulting schedule after auto-scheduling is correct?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/halide/Halide/pull/712#issuecomment-88176981>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/712#issuecomment-88184662>.
>

steven-johnson commented:
It's definitely vector-specific, at least. +1 to passing it a Target.

abadams commented:
The test should check that the expected transformations have been applied.
One way to do this is to validate the lowered code. E.g. call
compile_to_module and write an IRVisitor that checks that an allocate node
exists for things that should be compute root, a for loop over a given
variable is indeed parallel, etc.

On Tue, Mar 31, 2015 at 11:09 AM, Andrew Adams <andrew.b.adams@gmail.com>
wrote:

> I guess auto_schedule could take a target - after all, the "correct"
> schedule is probably target-specific.
>
>
>
> On Tue, Mar 31, 2015 at 10:44 AM, Zalman Stern <notifications@github.com>
> wrote:
>
>> +1 to Steven's comment. 256 bit on AVX prior to AVX2 is a lose for integer
>> ops.
>>
>> -Z-
>>
>>
>> On Tue, Mar 31, 2015 at 10:19 AM, Tyler Denniston <
>> notifications@github.com>
>> wrote:
>>
>> > To clarify: should the tests in tests/correctness only validate the
>> output
>> > of a pipeline which has been auto-scheduled? Or should they validate
>> that
>> > the resulting schedule after auto-scheduling is correct?
>> >
>> > —
>> > Reply to this email directly or view it on GitHub
>> > <https://github.com/halide/Halide/pull/712#issuecomment-88176981>.
>>
>> >
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/halide/Halide/pull/712#issuecomment-88184662>.
>>
>
>

tdenniston commented:
`auto_schedule` now gets a Target. I also changed the vectorize heuristic to use `natural_vector_size` and added a test for that.

@abadams Is the way I set up the tests (just inspecting the schedule objects directly) all right, or should I reimplement using IRVisitor as you suggested?

abadams commented:
That looks better than what I was suggesting. For some reason I was thinking you couldn't introspect on the schedules from user code.

tdenniston commented:
Ok, the remaining outstanding thing for now is removing the schedule reset. My comment might have gotten lost above, but I'm not sure how to check for conflicting scheduling decisions in the auto scheduler. The only way that comes to mind is replicating `validate_schedule` (from Lower.cpp) to return an error instead of aborting. Then, I could call that function after every individual scheduling decision, and roll back the most recent decision on error. However, that approach seems quite slow. Is there a better solution than that?

abadams commented:
If the schedule conflicts, then it'll throw an error already, and the user
will realize that some scheduling they did before calling auto_schedule
isn't compatible with what auto_schedule is doing.

On Tue, Mar 31, 2015 at 11:47 AM, Tyler Denniston <notifications@github.com>
wrote:

> Ok, the remaining outstanding thing for now is removing the schedule
> reset. My comment might have gotten lost above, but I'm not sure how to
> check for conflicting scheduling decisions in the auto scheduler. The only
> way that comes to mind is replicating validate_schedule (from Lower.cpp)
> to return an error instead of aborting. Then, I could call that function
> after every individual scheduling decision, and roll back the most recent
> decision on error. However, that approach seems quite slow. Is there a
> better solution than that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/712#issuecomment-88205500>.
>

abadams commented:
superceded by #1357
